src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +309,516 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
+
+	nonColocatedPlanningContext.anchorRelationRestrictionList =
+		anchorRelationRestrictionContext->relationRestrictionList;
+	nonColocatedPlanningContext.anchorAttributeEquivalances =
+		anchorRestrictionEquivalences;
+
+	/* handle from clause subqueries first */
+	RecursivelyPlanNonColocatedJoinWalker((Node *) joinTree, &nonColocatedPlanningContext,
+										  context);
+
+	/* handle subqueries in WHERE clause */
+	RecursivelyPlanNonColocatedSubqueriesInWhere(subquery, &nonColocatedPlanningContext,
+												 context);
+}
+
+
+/*
+ * AnchorSubqueryRte gets a query and searches for a subquery within the join tree of
+ * the query such that we can use it as our anchor subquery during our planning.
+ *
+ * The function returns NULL if it cannot find a proper range table entry for our
+ * purposes.
+ */
+static Query *
+AnchorSubquery(Query *subquery)
+{
+	FromExpr *joinTree = subquery->jointree;
+	Relids joinRelIds = get_relids_in_jointree((Node *) joinTree, false);
+	int currentRTEIndex = -1;
+	RangeTblEntry *rteRelation = NULL;
+
+	/*
+	 * Pick a random anchor subquery (i.e., the first) for now. We might consider
+	 * picking a better rte as the anchor. For example, we could iterate on the
+	 * joinRelIds, and check which rteIndex has more distribution key equiality
+	 * with rteIndexes. For the time being, the current primitive approach helps
+	 * us in many cases.
+	 */
+	while ((currentRTEIndex = bms_next_member(joinRelIds, currentRTEIndex)) >= 0)
+	{
+		RangeTblEntry *rte = rt_fetch(currentRTEIndex, subquery->rtable);
+
+		/*
+		 * We always prefer subqueries if we can find a suitable one for our
+		 * purposes.
+		 *
+		 * For the subquery, make sure that our subquery contains at least one
+		 * distributed table and doesn't have a set operation.
+		 *
+		 * The set operation restriction might sound weird, but, the restriction
+		 * equivalance generation functions ignore set operations. We should
+		 * integrate the logic in SafeToPushdownUnionSubquery() to
+		 * GenerateAllAttributeEquivalences() such that the latter becomes aware of
+		 * the set operations.
+		 *
+		 * In case we cannot find a subquery, we pick a relation, and wrap it into
+		 * a subquery.
+		 */
+		if (rte->rtekind == RTE_SUBQUERY &&
+			QueryContainsDistributedTableRTE(rte->subquery) &&
+			rte->subquery->setOperations == NULL &&
+			!ContainsUnionSubquery(rte->subquery))
+		{
+			return rte->subquery;
+		}
+		else if (rteRelation == NULL && rte->rtekind == RTE_RELATION)
+		{
+			rteRelation = rte;
+		}
+	}
+
+	/* if we found a relation, wrap into a subquery */
+	if (rteRelation)
+	{
+		return WrapRteRelationIntoSubquery(rteRelation);
+	}
+
 	return NULL;
 }
 
 
+/*
+ * WrapRteRelationIntoSubquery wraps the given relation range table entry
+ * in a newly constructed "(SELECT * FROM table_name as anchor_relation)" query.
+ */
+static Query *
+WrapRteRelationIntoSubquery(RangeTblEntry *rteRelation)
```

Comment:
could this ever lead to the table being pulled to the coordinator? (e.g. at a higher level of recursion) if that's something we do, is there a case for pushing down filters and projections here?